### ISSUE TITLE
Setter is not being called during init

### ISSUE THREAD

**Issue by orfisko** [pydantic/pydantic#6696]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I have noticed that if a setter is defined on a computed_field/property, this setter is not being called during the initialization of the model when a value is being supplied for that field. I find that somewhat counterintuitive. I use computed fields not for computations but just to have more flexible attributes in my model which can assume different values depending on the context.

### Example Code

```Python
from pydantic import computed_field, BaseModel, ConfigDict

class TestModel(BaseModel):
    _x: int
    model_config = ConfigDict(extra='forbid')

    @computed_field
    @property
    def x(self) -> int:
        return self._x

    @x.setter
    def x(self, value):
        print("Setter is called")
        self._x = value


m = TestModel(x=1)
# pydantic_core._pydantic_core.ValidationError: 1 validation error for TestModel
# x
#   Extra inputs are not permitted [type=extra_forbidden, input_value=1, input_type=int]
#     For further information visit https://errors.pydantic.dev/2.1.2/v/extra_forbidden
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0.2
pydantic-core version: 2.1.2 release build profile
                 install path: /home/hendrik/PycharmProjects/hardwareservice/venv/lib/python3.10/site-packages/pydantic
python version: 3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0]
                     platform: Linux-5.19.0-46-generic-x86_64-with-glibc2.35
     optional deps. installed: ['email-validator', 'typing-extensions']
```

Selected Assignee: @hramezani

Selected Assignee: @davidhewitt

**Comment by hramezani** [comment]
Thanks @orfisko for this issue 🙏 

Question: What do you want to do exactly? Could you please explain more?

You can have fields that are not included in the output by excluding them in dump.

**Comment by hramezani** [comment]
Closing as haven't got a repsonse

**Comment by orfisko** [comment]
I think my expectation is clear. I expect the setter to be called during
init...

Op ma 24 jul. 2023 09:03 schreef Hasan Ramezani ***@***.***>:

**Comment by orfisko** [comment]
@hramezani can you please open up this issue again? Just closing it because you are not getting answers on a question which is superfluous is pretty lame. Just for the sake of clarity: I expect 'computed_field' to mimic the behavior of the @property decorator in python. So, once a setter is defined it should also be called during init and not solely after the init as is currently the case in Pydantic

**Comment by hramezani** [comment]
@orfisko sure. Based on the [doc](https://docs.pydantic.dev/latest/usage/computed_fields/), computed_field is used in serialization and compute based on other fields. So, I think setter should not work there.

But let's see what @samuelcolvin thinks

**Comment by orfisko** [comment]
Like I mentioned, computed field was for me the long awaited functionality of pydantic to support the native @property decorator in python. There are numerous questions on stackoverflow and even[ pypi project](https://pypi.org/project/pydantic-property/) to get 'property' like functionality in v1. Therefore I am inclined to think it makes for a lot of the users a big difference having this well implemented. As a sidenote,  I don't understand the choice for naming the decorator to cater for this 'computed_field', there is so much more to this than using it for computations. I use this 'computed field' to generate a technical label based on other attribute values. It is so much nicer to address this property as instance.label and have it natively being featured in model dumps than coming up with all sorts of workarounds.

**Comment by orfisko** [comment]
@davidhewitt would you have a moment to spare to reflect on this?

**Comment by davidhewitt** [comment]
What you ask is for is an interesting feature but I think the fundamental problem with attempting to validate and input data to a property `setter` is that there's no way we can guarantee the input does not conflict with other fields.

For example, what should we do in this case?

```python
class A(BaseModel):
    x: int

    @computed_field
    @property
    def y(self) -> int:
        return self.x + 1

    @y.setter
    def y(self, value):
        self.x = value - 1
        
A(x=1, y=3) 
```

**Comment by orfisko** [comment]
Frankly, I don't see the problem. Why would it need to fail validation? Validation was, as far as I understood it, related to whether or not the attributes contain the type that was mentioned in the basemodel/computed field as being the type of the attribute. Both values are still ints, so in that sense validation is still working according to me. Validation would need to look at the end result. If that end result is still in line with what was specified, than happy days (I am aware there is far more to it in pydantic than that, with before and after validators etc., but I presume you are catching my drift).
If someone decides to mess around with those ints when using setters (as you did ;)), than that's their responsibility. I would simply expect pydantic to mimic the default 'property' behaviour in python as much as possible. If I have a setter defined on a field, it should not complain to me that I am not able to set that value because no extra attributes are allowed. Reasons is simple: I am not setting an extra attribute. I just want it to call the setter in that case and get out of the way.
```python
class B:
    __slots__ = ('x')

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __repr__(self):
        return f"x={self.x} y={self.y}"

    @property
    def y(self) -> int:
        return self.x + 1

    @y.setter
    def y(self, value):
        self.x = value - 1

b1 = B(x=2, y=5)
print(b1) # x=4 y=5
```
Defining __slots__ with the attributes that can be set comes closest according to me to what pydantic does with the 'extra' config flag. By defining those slots in the case above, I also prevent it from creating additional attributes. Still, it does not yell at me about not being able to set an extra attribute because I address it in the init, simply because it finds a setter to call.

**Comment by Viicos** [comment]
@orfisko I'm afraid I'm having trouble understanding what you mean by:
> I expect 'computed_field' to mimic the behavior of the @Property decorator in python. So, once a setter is defined it should also be called during init and not solely after the init as is currently the case in Pydantic

In standard Python, you could achieve this with the following (altough not common):

```python
class A:
    def __init__(self, x: int) -> None:
        self.x = x

    @property
    def x(self) -> int:
        return self._x

    @x.setter
    def x(self, value: int) -> None:
        self._x = value
```

But Pydantic isn't working like normal classes, it follows the same principles as dataclasses, meaning the `__init__` is generated dynamically (kind of) based on the class annotations. So even in standard Python with dataclasses:

```python
from dataclasses import dataclass, field

@dataclass
class A:
    _x: int = field(init=False)

    @property
    def x(self) -> int:
        return self._x

    @x.setter
    def x(self, value: int) -> None:
        print("Setter is called")
        self._x = value

a = A()
# How could the instanciation of A calls x's setter?
```

Sure, there are workarounds (defining the `__init__` method explicitly, playing with `dataclasses.InitVar` and `__post_init__`, but be careful as it could lead to some [unwanted side effects](https://gist.github.com/Viicos/26d8cc7ade6fb8828448142e8bf0951b)).

Maybe you could achieve what you want this way:

```python
from pydantic import computed_field, BaseModel, ConfigDict


class TestModel(BaseModel):
    _x: int
    model_config = ConfigDict(extra='forbid')

    def __init__(self, x: int, **kwargs) -> None:
        super().__init__(**kwargs)
        self.x = x

    @computed_field
    @property
    def x(self) -> int:
        return self._x

    @x.setter
    def x(self, value):
        print("Setter is called")
        self._x = value
```

**Comment by orfisko** [comment]
@Viicos thanks for your last suggestion. I have also considered the last option you mentioned, and it would indeed solve my issue.

I would agree that if you use pydantic's dataclass it should mimic the default python dataclass behaviour as much as possible. If it appears that in a python dataclass the setter is also ignored in case you would call `a=A(x=1)`, I could live with that. 
Note however the difference between `a =A(x=1)` and calling just `A()` in your example: I am not fighting the case of calling the setter in the latter case, that would not make any sense. 

Nevertheless, in case of defining a model when inheriting a BaseModel class, I remain convinced that it should work as I mentioned without needing to overwrite the __init__ . It would also allow to solve the logical inconsistency faced when trying to set a property with a setter in the model when extra="forbid" is configured. Whichever way you look at it, that does not make any sense.

**Comment by Viicos** [comment]
Unfortunately there's no way to tell to type checkers that `x` is to be included in the `__init__` method. Pydantic uses [`dataclass_transform`](https://docs.python.org/3/library/typing.html#typing.dataclass_transform) for this purpose, and iirc there's no way to indicate that property setters should be included.

My humble opinion is I don't think this is a good idea: it doesn't follow how it behaves with standard Python (specifically dataclasses). Sure Pydantic can deviate from the dataclass behavior, but the fact that `dataclass_transform` won't support it makes me think it will make things more confusing.

But I do understand your arguments as well.

**Comment by orfisko** [comment]
Well, I think my arguments make sense in particular because pydantic offers two ways to create a class with validated attributes: using the dataclass decorator and subclassing BaseModel. Not including this functionality in the former is fine to prevent any confusion, for the latter that is a much harder case to make imho.

In any case, I am willing to beef up the computed_field/property documentation with information and examples from this and the other issue we discussed about properties. The current documentation is very limited, I think it merits some more space. It would also be a nice way for me to do something back without needing to be as technically savvy as you guys :).

**Comment by gsakkis** [comment]
+1 for this feature request based on a [similar use case](https://github.com/pydantic/pydantic/discussions/7127#discussioncomment-6739351).

**Comment by dmontagu** [comment]
It seems to me that this feature request is making the assumption that passing a keyword argument to the `__init__` should "obviously" function where the `__init__` calls `setattr(self, key, value)`. But I just don't think that's the case, the following seems just as reasonable to me in the abstract:


```python
class A:
    def __init__(self, x: int) -> None:
        self._x = x

    @property
    def x(self) -> int:
        return self._x

    @x.setter
    def x(self, value: int) -> None:
        print("Setter is called")
        self._x = value
```

In particular, I think it would be a breaking change to make setters start getting called on calls to `__init__`, since I'm sure there is existing code out there using it and relying on the setter _not_ being called by the `__init__`.

While it's easier to argue that few people are probably using `@computed_field` in this way, I think ultimately it's more complex to implement _and_ I think it's not clear to me that the setter should be called by the `__init__`. (I personally would argue that that's less intuitive, not more.)

**Comment by gsakkis** [comment]
I have trouble coming up with a (non-pathological) use case where calling a setter in `__init__` would be a breaking change. The whole idea of Python properties is that from a user point of view they should be indistinguishable from attributes, otherwise we'd use explicit getter and setter methods.

To go back to the original example, as things stand `TestModel(x=1)` raises either a `ValidationError` (if `extra="forbid"`) or an `AttributeError` as soon as the `x` attribute is accessed (if `extra="ignore"`). If `TestModel(x=1)` is to be allowed, the only reasonable semantics could have by default  is `self.x = x`. Anything else (e.g. `self._x = x`) is custom behavior that has to be explicitly spelled out in overriding `__init__`. This issue is about making `__init__` treat computed fields the same way as regular (non-computed) fields. Honestly it looks more like a bug fix than a feature request to me.

**Comment by gsakkis** [comment]
Nice article, the fact that the (only) working solution works by having the setter called implicitly in `__init__` is a further argument for this issue. By "bit of magic" I assume you mean the `field(init=False, repr=False)` declaration for the private attribute but guess what, you can omit the whole declaration and it still works:

```py
@dataclass
class Vehicle:

    wheels: int

    @property
    def wheels(self) -> int:
        print("getting wheels")
        return self._wheels

    @wheels.setter
    def wheels(self, wheels: int):
        print("setting wheels to", wheels)
        self._wheels = wheels
```

**Comment by samuelcolvin** [comment]
We won't call setters on init because we want to build a dict, then set `__dict__` for performance reasons.

If you want to customize the model, use field validators, model validators, `__pydantic_post_init__` or computed fields.

**Comment by gsakkis** [comment]
> If you want to customize the model, use field validators, model validators, __pydantic_post_init__ or computed fields.

An example of how any of these could make `TestModel(x=1)` work would be great. AFAICT [overriding `__init__`](https://github.com/pydantic/pydantic/issues/6696#issuecomment-1677951385) seems the only workaround.

**Comment by decision-dev** [comment]
If it's still helpful, this is how I address this issue:
Yes, it does override __init__
It also assumes that you are ok with validate_assignment=True 

```
from typing import Any, Annotated

from pydantic import BaseModel, PrivateAttr, computed_field, validate_call


class MyBaseModel(BaseModel, validate_assignment=True):                             # this solution will validate on assignemnt for the computed_fields, so this makes everythign consistent.
    """BaseModel that adds initialization for computed fields"""

    def __init__(__pydantic_self__, **data: Any) -> None:  # pyright: ignore[reportSelfClsParameterName]  # to be consistent with pydantic.BaseModel.__init__ which uses __pydantic_self__
        super().__init__(**data)
        for field in __pydantic_self__.model_computed_fields:                       # for all of the computed fields
            if field in data:                                                       # if a value was passed to initialize the property
                if not getattr(__pydantic_self__.__class__, field).fset:            # if the property does not have a setter
                    raise TypeError(f"Can't initialize {field}. It is read only.")  # raise an error (You might want to make this a ValidationError)
                setattr(__pydantic_self__, field, data[field])                      # otherwise, the property does hava setter, so use it with the passed value


class WithProperties(MyBaseModel):
    """Example of a simple class with computed_fields"""

    x: int                                                                          # normal field
    _y: Annotated[str, PrivateAttr()] = ''                                          # private field to be accessed via property y

    @computed_field
    @property
    def y(self) -> str:                                                             # property y getter
        return self._y                                                              # Note: if self._y does not have a default value, the getter has to handle not hasattr(self, '_y')

    @y.setter
    @validate_call                                                                  # including @validate_call in all of your setters provides validation
    def y(self, value: str) -> None:                                                # property y setter
        self._y = value.upper()

    @computed_field
    @property
    def z(self) -> str:                                                             # read only property z
        return f"{self.x} {self.y}"

**Comment by rocketengineer1982** [comment]
I've encountered a related problem in #12159 .  I have computed fields with associated setters and I need those setters to be called when creating an instance of the model using `model_validate`.  I've found a work-around using a custom base class that during `__init__` checks for computed fields that have setters, and calls those setters if values are provided in the keyword arguments passed to `__init__`.  It validates the values it is setting to the computed fields using a dynamically generated pydantic model.
